state-chain/pallets/cf-chain-tracking/src/migrations/v2.rs

```diff
@@ -0,0 +1,141 @@
+				CurrentChainState::<T, I>::translate(|old| old.map(v1::FromV1::from_v1))
+					// XXX: is it okay to panic here? How to signalise an error otherwise?
+					.expect("failed to decode v1-storage");
+			if let Some(translated) = translated_opt {
+				CurrentChainState::<T, I>::put(translated);
+			}
+		}
+		// For the chains other than Bitcoin `v1::FromV1` should be defined as
+		//    something that explodes rather than silently corrupts the data.
+
+		Weight::zero()
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<Vec<u8>, DispatchError> {
+		Ok(Default::default())
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn post_upgrade(_state: Vec<u8>) -> Result<(), DispatchError> {
+		Ok(())
+	}
+}
+
+mod v1 {
+	use crate::*;
+
+	pub trait FromV1 {
+		type OldType: Decode;
+		fn from_v1(old: Self::OldType) -> Self;
+	}
+
+	#[derive(Copy, Clone, RuntimeDebug, PartialEq, Eq, Encode, Decode, MaxEncodedLen, TypeInfo)]
+	pub enum Never {}
+
+	macro_rules! impl_unreachable_from_v1_for_chain {
+		($chain: ty) => {
+			impl FromV1 for crate::ChainState<$chain> {
+				type OldType = Never;
+				fn from_v1(_: Self::OldType) -> Self {
+					unreachable!(
+						"We are not supposed to have an instance of {}",
+						core::any::type_name::<Self::OldType>()
+					)
+				}
+			}
+		};
+	}
+	impl_unreachable_from_v1_for_chain!(cf_chains::Ethereum);
+	impl_unreachable_from_v1_for_chain!(cf_chains::Polkadot);
+
+	pub mod btc {
+		use cf_chains::btc::{BitcoinFeeInfo, BitcoinTrackedData};
+
+		use super::FromV1;
+		use crate::*;
+
+		// The following type-aliases and constants are defined here intentionally (as opposed to
+		// being imported). The types and values should correspond to the types and values that were
+		// in effect right before this migration.
+		pub type BtcBlockNumber = u64;
+		pub type BtcAmount = u64;
+
+		const BYTES_PER_KILOBYTE: BtcAmount = 1024;
+		const INPUT_UTXO_SIZE_IN_BYTES: BtcAmount = 178;
+		const OUTPUT_UTXO_SIZE_IN_BYTES: BtcAmount = 34;
+		const MINIMUM_BTC_TX_SIZE_IN_BYTES: BtcAmount = 12;
+
+		#[derive(
+			Copy, Clone, RuntimeDebug, PartialEq, Eq, Encode, Decode, MaxEncodedLen, TypeInfo,
+		)]
+		pub struct FeeInfo {
+			pub fee_per_input_utxo: BtcAmount,
+			pub fee_per_output_utxo: BtcAmount,
+			pub min_fee_required_per_tx: BtcAmount,
+		}
+
+		#[derive(
+			Copy, Clone, RuntimeDebug, PartialEq, Eq, Encode, Decode, MaxEncodedLen, TypeInfo,
+		)]
+		pub struct TrackedData {
+			pub block_height: BtcBlockNumber,
+			pub tracked_data: FeeInfo,
+		}
+
+		impl FromV1 for ChainState<cf_chains::Bitcoin> {
+			type OldType = TrackedData;
+
+			fn from_v1(TrackedData { block_height, tracked_data }: TrackedData) -> Self {
+				log::info!("upgrading {} @{:?}", core::any::type_name::<Self>(), block_height);
+
+				fn undo(derived_fee: u64, size: u64) -> BtcAmount {
+					let a = derived_fee * BYTES_PER_KILOBYTE;
+					let q = a / size;
+					let r = a % size;
+
+					assert!(r == 0 || a == BtcAmount::MAX);
```

Comment:
Confused about the `a == BtcAmount::MAX` case, why does this matter? I assume you're saying "there might be a remainder if the multiplication saturated" ? - (which ofc requires use of saturating_mul above). 

I definitely like the use of asserts in general - but I think in this kind of code - i.e. runtime code, it's best to write defensive code that can't panic (which is what you've done by doing the `max()` chaining for example). So I'd probably err on the side of not including the assert - and using testing / reviews to check this - in particular because it's not clear that it would matter if this condition didn't hold, and could have very bad consequences if the assert is hit - i.e. a panic in the runtime, which is quite difficult to recover from.